src/controllers/shared.controllers/director.residential.address.link.controller.ts

```diff
@@ -0,0 +1,123 @@
+import { HA_TO_SA_ERROR } from "../../utils/constants";
+import { CompanyAppointment } from "private-api-sdk-node/dist/services/company-appointments/types";
+import { getCompanyAppointmentFullRecord } from "../../services/company.appointments.service";
+import { compareAddress } from "../../utils/address";
+
+export const getResidentialLink = async (req: Request, res: Response, next: NextFunction, templateName: string, backUrlPath: string, isUpdate: boolean) => {
+  try {
+    const transactionId = urlUtils.getTransactionIdFromRequestParams(req);
+    const submissionId = urlUtils.getSubmissionIdFromRequestParams(req);
+    const session: Session = req.session as Session;
+
+    const officerFiling = await getOfficerFiling(session, transactionId, submissionId);
+    const directorName = await getDirectorNameBasedOnJourney(isUpdate, session, req, officerFiling);
+
+    return res.render(templateName, {
+      templateName: templateName,
+      backLinkUrl: urlUtils.getUrlToPath(backUrlPath, req),
+      directorName: formatTitleCase(directorName),
+      ha_to_sa: calculateHaToSaRadioFromFiling(officerFiling.isHomeAddressSameAsServiceAddress),
+    });
+  } catch (e) {
+    return next(e);
+  }
+};
+
+export const postResidentialLink = async (req: Request, res: Response, next: NextFunction, templateName: string, backUrlPath: string, isUpdate: boolean) => {
+  try {
+    const transactionId = urlUtils.getTransactionIdFromRequestParams(req);
+    const submissionId = urlUtils.getSubmissionIdFromRequestParams(req);
+    const session: Session = req.session as Session;
+    const isHomeAddressSameAsServiceAddress = calculateHaToSaBooleanValue(req);
+    
+    if (isHomeAddressSameAsServiceAddress === undefined) {
+      const linkError = createValidationErrorBasic(HA_TO_SA_ERROR, DirectorField.HA_TO_SA_RADIO);
+      const officerFiling = await getOfficerFiling(session, transactionId, submissionId);
+      const directorName = await getDirectorNameBasedOnJourney(isUpdate, session, req, officerFiling);
+      return res.render(templateName, {
+        templateName: templateName,
+        backLinkUrl: urlUtils.getUrlToPath(backUrlPath, req),
+        directorName: formatTitleCase(directorName),
+        errors: formatValidationErrors([linkError])
+      });
+    }
+
+    const officerFilingBody: OfficerFiling = {
+      isHomeAddressSameAsServiceAddress: isHomeAddressSameAsServiceAddress
+    };
+
+    if(isUpdate){
+      const officerFiling = await getOfficerFiling(session, transactionId, submissionId);
+      const appointmentId = officerFiling.referenceAppointmentId as string;
+      const companyNumber= urlUtils.getCompanyNumberFromRequestParams(req);
+      const companyAppointment: CompanyAppointment = await getCompanyAppointmentFullRecord(session, companyNumber, appointmentId);
+      if(isHomeAddressSameAsServiceAddress !== companyAppointment.residentialAddressIsSameAsServiceAddress){
```

Comment:
companyAppointment.residentialAddressIsSameAsServiceAddress can be undefined. See comparison in https://github.com/companieshouse/officer-filing-web/pull/330 Also officerFilingBody.residentialAddressHasBeenUpdated is not always set so maybe add a function and do:

officerFilingBody.residentialAddressHasBeenUpdated = checkIsResidentialAddressUpdated(officerFiling.residentialAddress, companyAppointment.usualResidentialAddress);